docs/porting_guides/porting_guide_v3.1.rst

```diff
@@ -0,0 +1,120 @@
+as a single entry:
+
+::
+
+    foo,foo>=1.0   # from collection A, B
+
+Now, with version 3.1, those dependencies are no longer combined and will appear in the combined Python requirements
+file as separate entries:
+
+::
+
+    foo  # from collection A
+    foo>=1.0  # from collection B
+
+If your container image has an older version of ``pip``, this change might cause an error during the image build
+process. This guide covers how to deal with this situation below.
+
+Common Issues
+=============
+
+This section lists some common errors that might be encountered when running ``ansible-builder`` version 3.1.
+
+ERROR: Double requirement given
+-------------------------------
+
+Because dependency sanitization has been removed, *all* Python requirements from included collections and user
+requirement files are passed along to the ``pip`` command that installs those requirements. Due to this change,
+the image build may exit abnormally with an error similar to the following:
+
+::
+
+    ERROR: Double requirement given: netaddr>=0.10.1 (from -r /tmp/src/requirements.txt (line 13)) (already in netaddr (from -r /tmp/src/requirements.txt (line 4)), name='netaddr')
+    Error: building at STEP "RUN /output/scripts/assemble": while running runtime: exit status 1
+
+This error comes from ``pip`` within one of the intermediate container images when attempting to install the Python
+requirements from included collections and/or from the user supplied requirements. This intermediate image is based
+on the base image defined within the Execution Environment file, and it means that the version of ``pip`` installed
+within that image is too old to handle duplicate requirement entries.
+
+To determine where the duplicate requirements are coming from, run ``ansible-builder`` with the ``--v3`` option
+to get more verbose output, then look for the output from the introspection phase. It will look similar to below:
+
+::
+
+    [3/4] STEP 12/13: RUN $PYCMD /output/scripts/introspect.py introspect --write-bindep=/tmp/src/bindep.txt --write-pip=/tmp/src/requirements.txt
+    Creating parent directory for /tmp/src/requirements.txt
+    ---
+    python:
+    - 'netaddr  # from collection ansible.netcommon'
+    - 'netaddr>=0.10.1  # from collection ansible.utils'
+
+In the example output above, the double Python requirement for ``netaddr`` is coming from the collections
+``ansible.netcommon`` and ``ansible.utils``.
+
+The solution requires upgrading ``pip`` within the base image to a version that contains an updated dependency resolver.
+Beginning with ``pip`` version 20.3, the dependency resolver can handle duplicate requirements whose versions do not
+conflict, so that version is the minimum required. Upgrade ``pip`` in the base image from within the Execution
+Environment file by adding these lines to it:
+
+.. code-block:: yaml
+
+    additional_build_steps:
+      append_base:
+        - RUN pip3 install -U pip
```

Comment:
Should we make this `RUN $PYCMD -m pip install -U pip`?  To avoid issues where the `python` specified and `pip3` are for different python versions? Is that available at this phase?

Reply:
Yes, great point. Definitely available in the `append_base` phase.